pkg/abaputils/abaputils.go

```diff
@@ -0,0 +1,254 @@
+
+	var connectionDetails ConnectionDetailsHTTP
+	var error error
+
+	if options.Host != "" {
+		// Host, User and Password are directly provided -> check for host prefix (double https)
+		match, err := regexp.MatchString(`^[hH][tT][tT][pP][sS]:\/\/.*`, options.Host)
+		if err != nil {
+			return connectionDetails, errors.Wrap(err, "Host prefix validation failed")
+		}
+		if match {
+			connectionDetails.URL = options.Host + oDataURL
+		} else {
+			connectionDetails.URL = "https://" + options.Host + oDataURL
+		}
+		connectionDetails.User = options.Username
+		connectionDetails.Password = options.Password
+	} else {
+		if options.CfAPIEndpoint == "" || options.CfOrg == "" || options.CfSpace == "" || options.CfServiceInstance == "" || options.CfServiceKeyName == "" {
+			var err = errors.New("Parameters missing. Please provide EITHER the Host of the ABAP server OR the Cloud Foundry ApiEndpoint, Organization, Space, Service Instance and a corresponding Service Key for the Communication Scenario SAP_COM_0510")
+			return connectionDetails, err
+		}
+		// Url, User and Password should be read from a cf service key
+		var abapServiceKey, error = ReadServiceKeyAbapEnvironment(options, c, cfLogoutOption)
+		if error != nil {
+			return connectionDetails, errors.Wrap(error, "Read service key failed")
+		}
+		connectionDetails.URL = abapServiceKey.URL + oDataURL
+		connectionDetails.User = abapServiceKey.Abap.Username
+		connectionDetails.Password = abapServiceKey.Abap.Password
+	}
+	return connectionDetails, error
+}
+
+// ReadServiceKeyAbapEnvironment from Cloud Foundry and returns it.
+// Depending on user/developer requirements if he wants to perform further Cloud Foundry actions
+// the cfLogoutOption parameters gives the option to logout after reading ABAP communication arrangement or not.
+func ReadServiceKeyAbapEnvironment(options AbapEnvironmentOptions, c command.ExecRunner, cfLogoutOption bool) (AbapServiceKey, error) {
+
+	var abapServiceKey AbapServiceKey
+	var serviceKeyBytes bytes.Buffer
+	var err error
+
+	//Logging into Cloud Foundry
+	config := cloudfoundry.LoginOptions{
+		CfAPIEndpoint: options.CfAPIEndpoint,
+		CfOrg:         options.CfOrg,
+		CfSpace:       options.CfSpace,
+		Username:      options.Username,
+		Password:      options.Password,
+	}
+
+	err = cloudfoundry.Login(config, c)
+
+	c.Stdout(&serviceKeyBytes)
+	if err == nil {
+		// Reading Service Key
+		log.Entry().WithField("cfServiceInstance", options.CfServiceInstance).WithField("cfServiceKey", options.CfServiceKeyName).Info("Read service key for service instance")
+
+		cfReadServiceKeyScript := []string{"service-key", options.CfServiceInstance, options.CfServiceKeyName}
+
+		err = c.RunExecutable("cf", cfReadServiceKeyScript...)
+
+	}
+	if err == nil {
+		var serviceKeyJSON string
+
+		if len(serviceKeyBytes.String()) > 0 {
+			var lines []string = strings.Split(serviceKeyBytes.String(), "\n")
+			serviceKeyJSON = strings.Join(lines[2:], "")
```

Comment:
[curiosity] what is contained in line 1?

Reply:
Die Rückgabe der cf cli sieht so aus:
```
Getting key 0510 for service instance abapGit as D070355...

{
 "abap": {
...
```